crates/api/src/arc.rs

```diff
@@ -0,0 +1,386 @@
+
+/// A basic definition of a storage arc compatible with the concept of
+/// storage and querying of items in a store that fall within that arc.
+#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq, Default)]
+pub enum BasicArc {
+    /// No DHT locations are contained within this arc.
+    #[default]
+    Empty,
+    /// All DHT locations are contained within this arc.
+    Full,
+    /// A specific range of DHT locations are contained within this arc.
+    ///
+    /// The lower bound is inclusive, the upper bound is exclusive.
+    Literal(u32, u32),
+}
+
+impl BasicArc {
+    /// Create a new arc.
+    ///
+    /// If both values are 0 then a full arc is returned.
+    /// Otherwise, equal values for start and end will result in an empty arc.
+    /// Otherwise, a literal arc is returned.
+    ///
+    /// It is valid to create a [BasicArc] directly when using the Full and Empty variants.
+    /// However, when creating an arc from bounds, it is recommended to use this function for
+    /// consistency.
+    pub fn new(start: u32, end: u32) -> Self {
+        if start == end {
+            if start == 0 {
+                BasicArc::Full
+            } else {
+                BasicArc::Empty
+            }
+        } else {
+            BasicArc::Literal(start, end)
+        }
+    }
+
+    /// Convert to a canonical form.
+    ///
+    /// This function will return a new arc that is equivalent to the original arc,
+    /// See the documentation for [BasicArc::new].
+    pub fn canonicalize(&self) -> Self {
+        match self {
+            BasicArc::Empty => BasicArc::Empty,
+            BasicArc::Full => BasicArc::Full,
+            BasicArc::Literal(start, end) => Self::new(*start, *end),
+        }
+    }
+
+    /// Get the min distance from a location to an arc in a wrapping u32 space.
+    /// This function will only return 0 if the location is covered by the arc.
+    /// This function will return u32::MAX if the arc is empty.
+    ///
+    /// All possible cases:
+    ///
+    /// ```text
+    /// s = arc_start
+    /// e = arc_end
+    /// l = location
+    ///
+    /// Arc wraps around, loc >= arc_start
+    ///
+    /// |----e-----------s--l--|
+    /// 0                      u32::MAX
+    ///
+    /// Arc wraps around, loc < arc_end
+    /// |-l--e-----------s-----|
+    /// 0                      u32::MAX
+    ///
+    /// Arc wraps around, loc outside of arc
+    /// |----e----l------s-----|
+    /// 0                      u32::MAX
+    ///
+    /// Arc does not wrap around, loc inside of arc
+    /// |---------s--l---e-----|
+    /// 0                      u32::MAX
+    ///
+    /// Arc does not wrap around, loc < arc_start
+    /// |-----l---s------e-----|
+    /// 0                      u32::MAX
+    ///
+    /// Arc does not wrap around, loc >= arc_end
+    /// |---------s------e--l--|
+    /// 0                      u32::MAX
+    /// ```
+    pub fn dist(&self, loc: u32) -> u32 {
```

Comment:
This has changed, because the distance to the upper bound needs to take into account the upper bound being exclusive. So with an upper bound of 4, a location of 5 is actually at distance 2.

Reply:
No longer changed